src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java

```diff
@@ -0,0 +1,151 @@
+import org.killbill.billing.plugin.api.invoice.PluginInvoiceItem;
+import org.killbill.billing.plugin.api.invoice.PluginInvoicePluginApi;
+import org.killbill.billing.util.callcontext.CallContext;
+import org.killbill.billing.util.callcontext.TenantContext;
+import org.killbill.clock.Clock;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+class HelloWorldInvoicePluginApi extends PluginInvoicePluginApi implements OSGIKillbillEventHandler {
+
+	public HelloWorldInvoicePluginApi(OSGIKillbillAPI killbillAPI, OSGIConfigPropertiesService configProperties,
+			Clock clock) {
+		super(killbillAPI, configProperties, clock);
+	}
+
+	/**
+	 * Returns additional invoice items to be added to invoice
+	 * <p>
+	 * This method produces two types of invoice items a. Tax Item on Invoice Item
+	 * b. Adjustment Item on only the very first Historical Invoice Tax Item
+	 * automatically ( just for Demo purpose )
+	 *
+	 * @param newInvoice The invoice that is being created.
+	 * @param dryRun     Whether it is dryRun or not
+	 * @param properties Any user-specified plugin properties, coming straight out
+	 *                   of the API request that has triggered this code to run.
+	 * @param callCtx    The context in which this code is running.
+	 * @return A new immutable list of new tax items, or adjustments on existing tax
+	 *         items.
+	 */
+	@Override
+	public List<InvoiceItem> getAdditionalInvoiceItems(final Invoice newInvoice, final boolean dryRun,
+			final Iterable<PluginProperty> properties, final CallContext callContext) {
+
+		UUID accountId = newInvoice.getAccountId();
+		Account account = getAccount(accountId, callContext);
+		Set<Invoice> allInvoices = getAllInvoicesOfAccount(account, newInvoice, callContext);
+		ImmutableList.Builder<InvoiceItem> additionalItems = ImmutableList.builder();
+
+		// Creating tax item for first Item of new Invoice
+		List<InvoiceItem> newInvoiceItems = newInvoice.getInvoiceItems();
+		InvoiceItem newInvoiceItem = newInvoiceItems.get(0);
+		BigDecimal charge = new BigDecimal("80");
+		InvoiceItem taxItem = PluginInvoiceItem.createTaxItem(newInvoiceItem, newInvoiceItem.getInvoiceId(),
+				newInvoice.getInvoiceDate(), null, charge, "Tax Item");
+		additionalItems.add(taxItem);
+
+		// Creating External Charge for first Item of new Invoice
+		InvoiceItem externalItem = PluginInvoiceItem.create(newInvoiceItem, newInvoiceItem.getInvoiceId(),
+				newInvoice.getInvoiceDate(), null, charge, "External Item", InvoiceItemType.EXTERNAL_CHARGE);
+		additionalItems.add(externalItem);
+
+		// Creating adjustment item for first Item of Historical Invoice
+		for (Invoice invoice : allInvoices) {
+			if (!invoice.getId().equals(newInvoice.getId())) {
+				List<InvoiceItem> invoiceItems = invoice.getInvoiceItems();
+				// Check for if any adjustment item exists for Historical Invoice
+				if (checkforAdjustmentItem(invoiceItems)) {
+					break;
```

Comment:
Is the `checkforAdjustmentItem` necessary? AFAIK, even if an adjustment item exists for an invoice item, you can still create another adjustment item.

Reply:
Yes. We can still create the adjustment item. Our intention was to give an overview of how adjustment item is created. And, as mentioned in the comment, we are adding adjustment item for only first historical invoice.